File: src/fukurou/runner/portablemc.py
"""バニラクライアントの起動に使う PortableMC の取得。"""

import hashlib
from pathlib import Path
import tarfile

from fukurou.net import download
from fukurou.runner.process import GameProcessError

# PortableMC はバージョンとチェックサムを固定し、改ざんされたバイナリを実行しないようにする
PORTABLEMC_VERSION = "5.0.4"
PORTABLEMC_ARCHIVE = f"portablemc-{PORTABLEMC_VERSION}-linux-x86_64-gnu.tar.gz"
PORTABLEMC_SHA256 = "b14d2dff5191dabf90414562820ffdddfb5ee1acf692729782b4691d55b7b4f8"
PORTABLEMC_URL = (
    f"https://github.com/theorzr/portablemc/releases/download/v{PORTABLEMC_VERSION}/{PORTABLEMC_ARCHIVE}"
)


def ensure_portablemc(tools_dir: Path) -> Path:
    """PortableMC をダウンロード・チェックサム検証・展開し、実行ファイルのパスを返す。"""
    executable = tools_dir / f"portablemc-{PORTABLEMC_VERSION}" / "portablemc"
    if executable.is_file():
        return executable
    archive = tools_dir / PORTABLEMC_ARCHIVE
    if not archive.is_file() or hashlib.sha256(archive.read_bytes()).hexdigest() != PORTABLEMC_SHA256:
        # download はチェックサムが一致しなければ保存せずに失敗する
        download(PORTABLEMC_URL, archive, sha256=PORTABLEMC_SHA256)
    with tarfile.open(archive) as tar:
        # アーカイブ内のディレクトリ構成に依存しないよう、実行ファイルだけを取り出す
        member = next((m for m in tar.getmembers() if m.isfile() and Path(m.name).name == "portablemc"), None)
        if member is None:
            raise GameProcessError("portablemc executable was not found in the archive")
        source = tar.extractfile(member)
        executable.parent.mkdir(parents=True, exist_ok=True)
        executable.write_bytes(source.read())
    executable.chmod(0o755)
    return executable

File: kotlin/src/main/kotlin/party/morino/fukurou/engine/client/PortableMc.kt
package party.morino.fukurou.engine.client

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.runInterruptible
import party.morino.fukurou.engine.net.CacheLock
import party.morino.fukurou.engine.net.Downloader
import party.morino.fukurou.error.SetupException
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.StandardCopyOption
import java.nio.file.attribute.PosixFilePermissions

/** バニラクライアントの起動に使う PortableMC の取得（runner/portablemc.py）。 */
internal object PortableMc {
    /** PortableMC はバージョンとチェックサムを固定し、改ざんされたバイナリを実行しないようにする。 */
    const val VERSION: String = "5.0.4"

    /** リリースのアーカイブ名。 */
    const val ARCHIVE: String = "portablemc-$VERSION-linux-x86_64-gnu.tar.gz"

    /** アーカイブの sha256。 */
    const val SHA256: String = "b14d2dff5191dabf90414562820ffdddfb5ee1acf692729782b4691d55b7b4f8"

    /** ダウンロード元。 */
    const val URL: String = "https://github.com/theorzr/portablemc/releases/download/v$VERSION/$ARCHIVE"

    /** アーカイブ内の実行ファイルの名前。 */
    private const val EXECUTABLE = "portablemc"

    /** toolsDir の下の実行ファイルの置き場所（<toolsDir>/portablemc-5.0.4/portablemc）。 */
    fun executable(toolsDir: Path): Path = toolsDir.resolve("portablemc-$VERSION").resolve(EXECUTABLE)

    /**
     * PortableMC をダウンロード・チェックサム検証・展開し、実行ファイルのパスを返す。
     * 同じ toolsDir を使う他のテストや JVM と重ならないよう、<toolsDir>/.portablemc-5.0.4 のロックの下で行う。
     */
    suspend fun ensure(toolsDir: Path, downloader: Downloader = Downloader()): Path {
        val executable = executable(toolsDir)
        return CacheLock.withLock(toolsDir.resolve(".portablemc-$VERSION")) {
            // 展開済みなら何もしない（置き換えは原子的なので、あれば完全なファイル）
            if (Files.isRegularFile(executable)) return@withLock executable
            // download はチェックサムが一致しなければ保存せずに失敗し、一致する既存のアーカイブはそのまま使う
            val archive = downloader.download(URL, toolsDir.resolve(ARCHIVE), SHA256)
            runInterruptible(Dispatchers.IO) { install(archive, executable) }
            executable
        }
    }

    /** アーカイブから実行ファイルだけを取り出し、755 にして原子的に置く。 */
    private fun install(archive: Path, executable: Path) {
        // アーカイブ内のディレクトリ構成に依存しないよう、実行ファイルだけを取り出す
        val bytes = Files.newInputStream(archive).use { TarGzReader.extract(it, EXECUTABLE) }
            ?: throw SetupException("portablemc executable was not found in the archive $archive")
        Files.createDirectories(executable.parent)
        val partial = executable.resolveSibling("$EXECUTABLE.part")
        try {
            Files.write(partial, bytes)
            Files.setPosixFilePermissions(partial, PosixFilePermissions.fromString("rwxr-xr-x"))
            // 途中まで書いたファイルを実行しないよう、書き終えてから置き換える
            Files.move(partial, executable, StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING)
        } finally {
            Files.deleteIfExists(partial)
        }
    }
}
